OpenRA.Mods.Common/Widgets/Logic/Editor/TileSelectorLogic.cs

```diff
@@ -18,52 +19,120 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 {
 	public class TileSelectorLogic : ChromeLogic
 	{
+		class TileSelectorTemplate
+		{
+			public readonly TerrainTemplateInfo Template;
+			public readonly string[] Categories;
+			public readonly string[] SearchTerms;
+			public readonly string Tooltip;
+
+			public TileSelectorTemplate(TerrainTemplateInfo template)
+			{
+				Template = template;
+				Categories = template.Categories;
+				Tooltip = template.Id.ToString();
+				SearchTerms = template.Categories.Append(Tooltip).ToArray();
+			}
+		}
+
+		readonly TileSet tileset;
+		readonly WorldRenderer worldRenderer;
 		readonly EditorViewportControllerWidget editor;
 		readonly ScrollPanelWidget panel;
 		readonly ScrollItemWidget itemTemplate;
+		readonly TileSelectorTemplate[] allTemplates;
+
+		string selectedCategory;
+		string searchFilter;
 
 		[ObjectCreator.UseCtor]
 		public TileSelectorLogic(Widget widget, WorldRenderer worldRenderer)
 		{
-			var rules = worldRenderer.World.Map.Rules;
-			var tileset = rules.TileSet;
+			tileset = worldRenderer.World.Map.Rules.TileSet;
+			this.worldRenderer = worldRenderer;
 
 			editor = widget.Parent.Get<EditorViewportControllerWidget>("MAP_EDITOR");
 			panel = widget.Get<ScrollPanelWidget>("TILETEMPLATE_LIST");
 			itemTemplate = panel.Get<ScrollItemWidget>("TILEPREVIEW_TEMPLATE");
 			panel.Layout = new GridLayout(panel);
 
-			var tileCategorySelector = widget.Get<DropDownButtonWidget>("TILE_CATEGORY");
-			var categories = tileset.EditorTemplateOrder;
+			allTemplates = tileset.Templates.Values.Select(t => new TileSelectorTemplate(t)).ToArray();
+
+			var orderedCategories = allTemplates.SelectMany(t => t.Categories)
+				.Distinct()
+				.OrderBy(CategoryOrder)
+				.ToArray();
+
+			Func<string, string> categoryTitle = s => s != null ? s : "Search Results";
 			Func<string, ScrollItemWidget, ScrollItemWidget> setupItem = (option, template) =>
 			{
 				var item = ScrollItemWidget.Setup(template,
-					() => tileCategorySelector.Text == option,
-					() => { tileCategorySelector.Text = option; IntializeTilePreview(widget, worldRenderer, tileset, option); });
+					() => selectedCategory == option,
+					() => { selectedCategory = option; InitializeTilePreview(); });
 
-				item.Get<LabelWidget>("LABEL").GetText = () => option;
+				var title = categoryTitle(option);
+				item.Get<LabelWidget>("LABEL").GetText = () => title;
 				return item;
 			};
 
+			var searchTextField = widget.GetOrNull<TextFieldWidget>("SEARCH_TEXTFIELD");
+			if (searchTextField != null)
+			{
+				searchTextField.OnTextEdited = () =>
+				{
+					searchFilter = searchTextField.Text.Trim();
+					selectedCategory = string.IsNullOrEmpty(searchFilter) ? orderedCategories.First() : null;
```

Comment:
An idea - to remove this line to return to original selected category, but it requires some changes where selectedCategory==null is checked.